### PR TITLE
Add new `driven(by:)` combiner

### PR DESCRIPTION
**Pull request by CarlEkman**
There is a recurring use case for plain void signals emitting values from other, readable signals or providers. One common example of this is:
```Swift
// Grab the latest value from the input field when button is pressed
bag += myButton
    .map { myTextField.value } // FIXME: Relies on outside state...
    .onValue { (input) in
        ...
}
```
This is not very reactive, and is more similar to imperative patterns. We do however have the `withLatestFrom` combiner:
```Swift
bag += myButton.providedSignal.readable(initial: ()) // FIXME: Hacky...
    .withLatestFrom(myTextField)
    .onValue { (_, input) in
        ...
}
```
However, in order for this to work _we need to make sure the signals are of the same kind_ – and making a button signal readable makes very little sense...

So how can we make plain void signals work better here? The intuition is that we should be able to have a plain signal `A` that fires `Void` events. Then we want to replace those emitted values with values from a read signal `B`:
```
----*------*------*-----*--| // A
    |      |      |     |
a)------b------c-----------| // B
    |      |      |     |
+--------------------------+
| withLatestFrom()         |
+--------------------------+
    |      |      |     |
----a------b------c-----c--|
```
This PR attempts to address this by making a convenience wrapper for signal providers with value `Void` and type `Plain`, that takes a readable signal and emits events from that (and not tuples).
```Swift
public extension SignalProvider where Value == Void, Kind.DropWrite == Plain {
    func withLatestFrom<S: SignalProvider>(_ other: S) -> CoreSignal<Plain, S.Value>
    where S.Kind.DropWrite == Read {
        return withLatestFrom(other.plain()).map { $1 }
    }
}
```

### CODE REVIEW

**Comment by CarlEkman** [comment]
Alternatively, since this is a bit different from the other `withLatestFrom` and doesn't return tuples, we might want to use another name, like `takeLatestFrom`.

Also, we might want this to not wrap `withLatestFrom` and use a direct implementation instead.

**Comment by andriichernenko** [comment]
This is a very common use case, so this function makes a lot of sense to me.

**Comment by andriichernenko** [comment]
> use another name, like `takeLatestFrom`

From the FRP point if view, we are doing essentially the same thing, so I don't see the need for another operator.

**Comment by mansbernhardt** [comment]
I like this, and think this addition makes sense. I think what you suggest above, might have been the intention of original addition in the PR https://github.com/iZettle/Flow/pull/30. But that it evolved during review to perhaps break the initial use case? 

Some feedback and questions....

Firstly I would assume your example:

```swift
bag += myButton.providedSignal.readable(initial: ()) // FIXME: Hacky...
    .withLatestFrom(myTextField)
    .onValue { (_, input) in
        ...
}
```

could be simplified to:

```swift
bag += myButton.readable()
    .withLatestFrom(myTextField)
    .onValue { (_, input) in
        ...
}
```

to make it more fair?

And as for your implementation:

```swift
public extension SignalProvider where Value == Void, Kind.DropWrite == Plain {
    func withLatestFrom<S: SignalProvider>(_ other: S) -> CoreSignal<Plain, S.Value>
    where S.Kind.DropWrite == Read {
        return withLatestFrom(other.plain()).map { $1 }
    }
}
```

Couldn't `Kind.DropWrite == Plain` just be `Kind == Plain`

But the bigger question is, what would happen if the button signals before other has ever been signaled? That value would not get through, would it?

And if we want that to work (which I think the use would expect), the returned signal should rather be a ReadSignal.

```swift
public extension SignalProvider where Kind == Plain, Value = () {
    func withLatestFrom<S: SignalProvider>(_ other: S) -> CoreSignal<Read, S.Value>
    where S.Kind.DropWrite == Read {
        return /// What goes here?
    }
}
```

and it will perhaps return:

```swift
return readable().withLatestFrom(other).map { $1 }
```

**Comment by CarlEkman** [comment]
Thanks for taking the time @mansbernhardt!

Yes, I forgot you can use `readable()`, but my point is still that making a `Signal<Void>` readable makes very little sense, and you still get a tuple of `(Void, T)` whereas I'd prefer `T`.

>Couldn't Kind.DropWrite == Plain just be Kind == Plain
Yes, that's better! 🙂 

Hmm, yes. Fair point - maybe it does make sense for it to return a readable signal. Nice!
Do you think it's okay to wrap the other `withLatestFrom` implementation, and simply dropping the first element of the tuple, or is it better to mimic the the other implementation but not use tuples at all?

**Comment by mansbernhardt** [comment]
I would say we skip the tuple and just return the second value as the first does not provide any additional value. This is similar as we did for the `with(weak:)` `()` overload:

```swift
public extension SignalProvider where Value == () {
    func with<T: AnyObject>(weak object: T) -> CoreSignal<Kind.DropWrite.DropWrite, T> {
        return with(weak: object).map { $1 }
    }
}
```

**Comment by andriichernenko** [comment]
> But the bigger question is, what would happen if the button signals before other has ever been signaled? That value would not get through, would it?

Would `other.atOnce().plain()` solve the problem? Something like 

```swift
public extension SignalProvider where Value == Void, Kind.DropWrite == Plain {
    func withLatestFrom<S: SignalProvider>(_ other: S) -> CoreSignal<Plain, S.Value>
    where S.Kind.DropWrite == Read {
        return withLatestFrom(other.atOnce().plain()).map { $1 }
    }
}
```

**Comment by mansbernhardt** [comment]
@andriichernenko It would probably solve it as well. But I think there are advantages to not loose the read part of `other`, even though I'm not sure it it will be of much use?

Perhaps it could be useful in a context such as:

```swift
let sentMessage = sendButton.withLatestFrom(sendText)
bag += sentMessage.atOnce().bindTo(vc, \.title) // Update title for every sent message
```

**Comment by mansbernhardt** [comment]
We might also consider to add some overloads for multiple others so instead of writing:

```swift
bag += loginButton.withLatestFrom(combineLatest(emailField, passwordField)).onValue { ... }
```

we could instead write:

```swift
bag += loginButton.withLatestFrom(emailField, passwordField).onValue { ... }
```

We might also consider if we should update readme.md  and the login sample project to use `withLatestFrom`?

**Comment by CarlEkman** [comment]
@mansbernhardt Yes I was also thinking about the `combineLatest` use case. Would be great if we could do that as well!

I agree that it would probably be useful to keep the readable property of the result.

**Comment by CarlEkman** [comment]
So, perhaps something like this?
```Swift
public extension SignalProvider where Value == Void, Kind == Plain {
    func withLatestFrom<S: SignalProvider>(_ other: S) -> CoreSignal<Read, S.Value>
        where S.Kind.DropWrite == Read {
        return readable().withLatestFrom(other).map { $1 }
    }

    func withLatestFrom<A: SignalProvider, B: SignalProvider>(_ a: A, b: B) -> CoreSignal<Read, (A.Value, B.Value)>
        where A.Kind.DropWrite == Read, B.Kind.DropWrite == Read {
        return readable().withLatestFrom(combineLatest(a, b)).map { $1 }
    }

    func withLatestFrom<A: SignalProvider, B: SignalProvider, C: SignalProvider>(_ a: A, b: B, c: C) -> CoreSignal<Read, (A.Value, B.Value, C.Value)>
        where A.Kind.DropWrite == Read, B.Kind.DropWrite == Read, C.Kind.DropWrite == Read {
            return readable().withLatestFrom(combineLatest(a, b, c)).map { $1 }
    }
}
```
Do we want to go all the way to `n`, just like `combineLatest`? 😄

File: Flow/Signal+Combiners.swift [hunk 1]
```diff
@@ -51,39 +51,79 @@ public func merge<Signals: Sequence>(_ signals: Signals) -> CoreSignal<Signals.I
...
+    ///
+    ///     ----*------*------*-----*--|
+    ///         |      |      |     |
+    ///     a)------b------c-----------|
+    ///         |      |      |     |
+    ///     +--------------------------+
+    ///     | withLatestFrom()         |
+    ///     +--------------------------+
+    ///         |      |      |     |
+    ///     ----a------b------c-----c--|
```
Comment (mansbernhardt) [inline Flow/Signal+Combiners.swift]: As we are now returning a ReadSignal this line should rather be `a)--a------b------c-----c--|` ?

File: Flow/Signal+Combiners.swift [hunk 2]
```diff
@@ -51,39 +51,79 @@ public func merge<Signals: Sequence>(_ signals: Signals) -> CoreSignal<Signals.I
...
+    func withLatestFrom<S: SignalProvider>(_ other: S) -> CoreSignal<Read, S.Value>
+        where S.Kind.DropWrite == Read {
+        return readable().withLatestFrom(other).map { $1 }
+    }
+
+    /// Returns a new readable signal returning the latest values from the given signals.
+    /// - Note: See `withLatestFrom(_:)` for more info.
+    func withLatestFrom<A: SignalProvider, B: SignalProvider>(_ a: A, _ b: B) -> CoreSignal<Read, (A.Value, B.Value)>
+        where A.Kind.DropWrite == Read, B.Kind.DropWrite == Read {
+        return readable().withLatestFrom(combineLatest(a, b)).map { $1 }
```
Comment (mansbernhardt) [inline Flow/Signal+Combiners.swift]: I would guess that 
```swift
return withLatestFrom(combineLatest(a, b))
```
would work as well? Use the initial non-tuple Void overload?

**Comment by mansbernhardt** [comment]
@CarlEkman it would be awesome if you took the time to look into if we could potentially update/simplify the readme and sample code as well.

File: README.md [hunk 3]
```diff
@@ -110,20 +110,22 @@ class LoginController: UIViewController {
...
-            // ...and retry the login request if the user chose to
-            self.showRetryAlert(for: error)
-          }.onValue { user in
-            // If login is successful, complete runLogin() with the user
-            completion(.success(user))
-          }
+      // If button is tapped, initiate potentially long running login request using input
+      bag += self.loginButton
+        .withLatestFrom(emailField, passwordField)
+        .onValue { (email, password) in
```
Comment (mansbernhardt) [inline README.md]: Are the brackets really required here? Would not `.onValue { email, password in` be accepted as well? I think the latter is easier and clearer for the eye if possible.

Comment (CarlEkman) [reply README.md]: Yes, it's just a style preference. The parentheses are not needed, it's just to show that these are "input parameters" captured by the following closure body. I can remove them 👍

**Comment by andriichernenko** [comment]
Standard library has overloads for the `==` operator for up to 6 arguments, maybe we should follow that example too.

**Comment by andriichernenko** [comment]
@mansbernhardt If we take button presses as an example, what does using the `.readable()` operator on it mean? Button presses are momentary, they don't have current value which we can read. I mean, yes, it is possible to achieve technically with Flow, but it just doesn't make sense. But that may just be me.

The use case you mentioned (with updating the title when sending a message), I would implement it this way:

```swift
sendButton
    .withLatestFrom(sendText) // Signal<String>
    .readable(initial: "some initial value")
    .atOnce()
    .bindTo(vc, \.title)
```
This might be more verbose, but it is less confusing in my opinion.

**Comment by mansbernhardt** [comment]
@andriichernenko 

Flow is a bit different from many other reactive frameworks to specify and guarantee initial values via the type system instead by documentation/convention. To maintain this type information such as readability through signal transforms, puts some limits on how different signal can be combined. As for your example, if you have an initial value different from what the text field currently has, yes then you can write it as above, or perhaps even clearer as:

```swift
sendButton
    .withLatestFrom(someTextField) // Signal<String>
    .start(with: "some initial value")
    .bindTo(vc, \.title)
```

But it's really common that the initial value is the same as the current value of the field, hence your code above would have to repeat the "some initial value" part:

```swift
let someTextField = UITextField(value: "some initial value")
sendButton
    .withLatestFrom(someTextField) // Signal<String>
    .start(with: "some initial value") // <<--- "some initial value" is repeated here
    .bindTo(vc, \.title)
```

Whereas if you let `withLatestFrom` return a readable signal you can avoid that:

```swift
let someTextField = UITextField(value: "some initial value")
sendButton
    .withLatestFrom(someTextField) // ReadSignal<String>
    .atOnce() // <-- no repetition of initial value
    .bindTo(vc, \.title)
```

I think the "Accessing initial state" in the [article](https://medium.com/izettle-engineering/expanding-on-signals-ad25daee4d64) showcases this in a nice way:

**Comment by andriichernenko** [comment]
@mansbernhardt okay, I see. I still can't get over the fact that a button push signal is readable though. What if we introduced another operator for `ReadSignal` that addresses this use case? Something like

```swift
let someTextField = UITextField(value: "some initial value")
someTextField
    .drivenBy(sendButton)
    .atOnce()
    .bindTo(vc, \.title)
```

**Comment by CarlEkman** [comment]
@andriichernenko I actually had a similar idea at first, something like:
```swift
let mySignal = myTextField.emitter(myButton)
```

**Comment by mansbernhardt** [comment]
@andriichernenko Sure, I don't really have any preference for spelling this `withLatestFrom`. I think we could spell this differently. To me, I think that the ordering of things in your example might be bit clearer as well, withLatestFrom feels a bit backwards. Then we could probably skip the many values overloads as well and simplify things:

```swift
bag += combineLatest(emailField, passwordField).drivenBy(sendButton).onValue { ... }
```

Not sure how we should name this and how other reactive frameworks name similar transforms.

Use `drivenBy` would perhaps also make it easier to reuse transforms:

```swift
let nameAndPassword = combineLatest(emailField, passwordField)
bag += nameAndPassword.map(isValid).bindTo(sendButton, \.isEnabled)
bag += nameAndPassword.drivenBy(sendButton).onValue(login)
```

**Comment by CarlEkman** [comment]
@mansbernhardt @andriichernenko I added a preliminary implementation of `driven(by:)`, please have a look!

I suppose we could remove the other overloads now then?

**Comment by mansbernhardt** [comment]
@CarlEkman I think we should go with `driven(by:)` instead and remove the new `withLatestFrom` variants, as the former is simpler and reads better. Make sure to also update the readme and sample projects.

**Comment by CarlEkman** [comment]
@mansbernhardt I started looking into it, but it didn't build and I had to start working on other things. I don't think it contained any references to `withLatestFrom` at least, so we don't have to change the API usage. I think we can do it in a later PR. I still don't have permission to merge this though.

**Comment by mansbernhardt** [comment]
@CarlEkman Ok, I can fix it in a follow up PR. If you could rebase this one, I will merge it.

**Comment by CarlEkman** [comment]
@mansbernhardt There seems to be some failing tests, but they seem like flaky stress tests to me... I don't think my changes have dirties anything there.

**Pull request merged**